Reintegrate KK with our fork of Kruise Kontrol with a remote API.

ahicks92 opened the pull request:
The fork is at https://github.com/factorio-access/kk-remote

See remote.lua over there for the API (with documentation!).

This also includes some fixes, since I basically rewrote it anyway.  And comments.  Please check the accuracy of the comments.

basically what I did was add a remote API matching our needs.  It can tell you whether it's active and whatever else.  The strings changed a bit, in particular the walking string is now "moved to (x, y)" literally with the parens.

Of course there's a bunch of open questions.  Firstly before continuing note that this is behind the ruler change by a commit, I'll rebase before landing.  The rulers are too annoying, I can't actually play with them yet.  At least fixing sounds is next on my list.

Anyway, a summary:

- Remove all the references to Kruise Kontrol global state in the mod that I spotted. Maybe some left.
- Get rid of all the guesswork and call the API.
- Now it's ctrl+alt+], with changes:
  - We don't allow blueprints to be in the hand, as those seem to break it badly. For now we just flat out error and make the player decide what to do as it's definitely less than clear.
  - We turn cursor mode off.  This also can freak KK out in a whole lot of ways
  - And of course we check the hand stack to see if it's a blueprint or book.
- the strings for KK are now localized (but let's discuss that elsewhere: `access.long-name` is annoying, can we just switch to `fa` before going hard in that direction?)

And now for the concerns, weirdness, and general questions.

- Didn't we do something about get_selected_ent_deprecated?  Thought we were moving to a new API.
- What are we going to do about inputs?  I'd rather review this early but the underlying issue is that it turns out we can't turn inputs off or disable them without breaking KK.  I think we will need to come up with a dummy key and move inputs we want to disable there.  I nominate ctrl+alt+shift+backspace, but let's discuss if we feel it's needed.
- A player's cursor_pos is very weird.  It seems like it's including fractional components except when it isn't.  Floor here seems to match right but...
- Are my changes in control.lua correct?  That seems like the kind of thing which is going to have a bug.
- Can we avoid toggling remote view off?  If this works with remote view on there is no  need in theory but in practice...
- It looks like all our alt inputs don't work except that they do?  Kinda unrelated but appareently factorio wants "CONTROLL", not "CTRL".  I haven't tested.  Mine didn't seem to work till I fixed it but there was a lot going wrong and it could just be human error.

Anyway I've done minimal testing.  I'm thinking we don't land for like a week.  It seems to work but this is a large change.  To test it yourself remove the kk zip and clone the other repo directly to (case sensitive) Kruise_Kontrol_Updated.

I'll make issues or just fix them, but I noticed three things that should be followed up on:

- We don't need to always announce KK state now that we have reliable reporting.  Let's consider how we want to handle that. I think exponential backoff or replacing it with a KK working sound.
- How we schedule around ticks happens in multiple ways and in addition doesn't use functions as values and one of the methods even relies on the function being global. (see `schedule` in control.lua).  There are much simpler, better ways which can balance work without inventing some new random modulus and shoving it somewhere.  easy to fold this into my pending event handling rework.
- One of the bigger cases around magic numbers instead of named constants turns out to be the walking mode.  This is basically incomprehensible unless you have previously memorized the indices of the modes.  We can fix this with a LuaLS enum table.
- Finally why can't we search based off just a position?  Probably something @LevFendi knows but if you don't have a bounding box and search a surface off cursor_pos which I thought was fractional you don't get any entities.

ahicks92 committed:
Reintegrate KK with our fork of Kruise Kontrol with a remote API.

LevFendi commented on data.lua:
```diff
@@ -2049,4 +2049,19 @@ data:extend({
+   {
+      name = "fa-kk-cancel",
+      type = "custom-input",
+      linked_game_control = "toggle-driving",
```
Why is it intentionally linked to driving? The ENTER key was selected for convenience.

LevFendi replied:
Because of the linked game control, perhaps it might be that the cancel operation does not fire properly unless you are next to a vehicle.

LevFendi commented on scripts/kruise-kontrol-wrapper.lua:
```diff
@@ -1,244 +1,115 @@
-end
+      -- We screwed around with the running modifier. Put it back based on
+      -- cursor mode.
+      fix_walk(pindex)
```
Info: The thing about messing with walking is that telestep mode goes bonkers in multiplayer, so in this walking mode we intentionally force the player walking speed to become zero. 
The new version of this file handles this matter correctly.

LevFendi commented on scripts/kruise-kontrol-wrapper.lua:
```diff
@@ -1,244 +1,115 @@
-end
+      -- If remote view was on turn it off.  unclear if this is necessary now
+      -- that KK isn't moving the mouse to work but it really doesn't hurt.
+      toggle_remote_view(pindex, false, true)
```
We will need to test if we can remove this but I prefer to.

LevFendi reviewed (commented):
The changes in the code all seem to be good. I will test before final approval.

ahicks92 committed:
More fixes, still doesn't announce cancellation

ahicks92 committed:
Also announce when kk is finished

ahicks92 committed:
Readme improvements, drop KK input modifications

ahicks92 committed:
Forgot about combat

ahicks92 committed:
Update the dep

LevFendi committed:
Revised KK in README.md

LevFendi committed:
Update comment in data-updates.lua

LevFendi committed:
List other KK versions as incompatible

LevFendi reviewed (approved):
Review complete, changes made in line comments, README, and info.json. Testing comes next...

LevFendi committed:
Link KK section to wiki page

LevFendi committed:
Fix valid character and cursor stack check

LevFendi committed:
Force cursor off is now silent ...

Toggling cursor mode and remote view can now be muted, while they are not muted by default.

LevFendi committed:
No more local fn to mute "force cursor off"

LevFendi committed:
StyLua (automation failed for some reason)

LevFendi committed:
Name the walking mode constants
